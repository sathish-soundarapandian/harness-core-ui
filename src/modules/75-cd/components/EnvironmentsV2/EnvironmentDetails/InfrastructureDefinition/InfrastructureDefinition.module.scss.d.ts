/* eslint-disable */
/**
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 **/
// this is an auto-generated file, do not update this manually
declare const styles: {
  readonly actionConfirmationDialogWrapper: string
  readonly body: string
  readonly buttonWrapper: string
  readonly closeButton: string
  readonly dialogStyles: string
  readonly infraDefinitionDetailsDrawer: string
  readonly infraDetailsConfigurationTab: string
  readonly infraDetailsReferencedByTab: string
  readonly modalFooter: string
  readonly nameIdCard: string
  readonly referenceByPageBody: string
  readonly tagRender: string
  readonly tagsPopover: string
  readonly yamlBuilder: string
}
export default styles
