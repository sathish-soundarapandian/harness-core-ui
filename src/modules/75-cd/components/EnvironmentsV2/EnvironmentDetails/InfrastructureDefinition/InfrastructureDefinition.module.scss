/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.tagsPopover {
  p {
    color: var(--grey-600);
  }
}

.dialogStyles {
  :global {
    .ModalDialog--body {
      padding: 0 !important;
    }
  }

  overflow: hidden;
}

.modalFooter {
  position: sticky;
  bottom: 0;
  background-color: var(--white) !important;
  z-index: 11;
  border-radius: 6px !important;
}

.nameIdCard {
  width: 40%;
  margin-bottom: var(--spacing-large);
}

.yamlBuilder {
  position: relative;

  .buttonWrapper {
    position: absolute;
    top: var(--spacing-medium);
    right: var(--spacing-large);
    display: flex;
    column-gap: var(--spacing-large);
    align-items: center;
    z-index: 2;
  }
}
