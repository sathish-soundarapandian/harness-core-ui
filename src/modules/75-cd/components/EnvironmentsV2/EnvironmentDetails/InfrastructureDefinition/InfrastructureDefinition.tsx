/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'

import { ButtonSize, ButtonVariation, Container, ModalDialog, Page } from '@harness/uicore'
import { useModalHook } from '@harness/use-modal'

import { useGetInfrastructureList } from 'services/cd-ng'
import { useStrings } from 'framework/strings'

import type { EnvironmentPathProps, ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { ContainerSpinner } from '@common/components/ContainerSpinner/ContainerSpinner'

import useRBACError from '@rbac/utils/useRBACError/useRBACError'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import RbacButton from '@rbac/components/Button/Button'

import { useTemplateSelector } from 'framework/Templates/TemplateSelectorContext/useTemplateSelector'
import InfrastructureList from './InfrastructureList/InfrastructureList'
import InfrastructureModal from './InfrastructureModal'

import css from './InfrastructureDefinition.module.scss'

export default function InfrastructureDefinition(): JSX.Element {
  const { accountId, orgIdentifier, projectIdentifier, environmentIdentifier } = useParams<
    ProjectPathProps & EnvironmentPathProps
  >()
  const { getString } = useStrings()
  const { getRBACErrorMessage } = useRBACError()
  const [selectedInfrastructure, setSelectedInfrastructure] = useState<string>('')

  const { data, loading, error, refetch } = useGetInfrastructureList({
    queryParams: {
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier,
      environmentIdentifier
    }
  })

  const { getTemplate } = useTemplateSelector()
  const onClose = (): void => {
    setSelectedInfrastructure('')
    hideModal()
  }

  const [showModal, hideModal] = useModalHook(
    () => (
      <ModalDialog
        isOpen
        isCloseButtonShown
        canEscapeKeyClose
        canOutsideClickClose
        enforceFocus={false}
        onClose={onClose}
        title={selectedInfrastructure ? getString('cd.infrastructure.edit') : getString('cd.infrastructure.createNew')}
        width={1128}
        height={840}
        className={css.dialogStyles}
      >
        <InfrastructureModal
          hideModal={onClose}
          refetch={refetch}
          environmentIdentifier={environmentIdentifier}
          selectedInfrastructure={selectedInfrastructure}
          getTemplate={getTemplate}
        />
      </ModalDialog>
    ),
    [refetch, selectedInfrastructure, setSelectedInfrastructure]
  )

  useEffect(() => {
    if (selectedInfrastructure) {
      showModal()
    }
  }, [selectedInfrastructure, showModal])

  return (
    <Container padding={{ left: 'medium', right: 'medium' }}>
      {loading ? (
        <ContainerSpinner />
      ) : error ? (
        <Page.Error>{getRBACErrorMessage(error)}</Page.Error>
      ) : (
        <>
          <RbacButton
            text={getString('pipelineSteps.deploy.infrastructure.infraDefinition')}
            font={{ weight: 'bold' }}
            icon="plus"
            onClick={showModal}
            size={ButtonSize.SMALL}
            variation={ButtonVariation.LINK}
            permission={{
              resource: {
                resourceType: ResourceType.ENVIRONMENT
              },
              permission: PermissionIdentifier.EDIT_ENVIRONMENT
            }}
          />
          <InfrastructureList
            list={data?.data?.content}
            showModal={showModal}
            refetch={refetch}
            setSelectedInfrastructure={setSelectedInfrastructure}
          />
        </>
      )}
    </Container>
  )
}
