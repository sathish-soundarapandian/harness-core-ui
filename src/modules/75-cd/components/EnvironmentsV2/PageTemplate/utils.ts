/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { useFeatureFlags } from '@common/hooks/useFeatureFlag'
import type { FilterProperties } from 'services/cd-ng'
import type { UseQueryParamsOptions } from '@common/hooks/useQueryParams';
import { useQueryParamsOptions } from '@common/hooks/useQueryParams'
import { COMMON_DEFAULT_PAGE_SIZE } from '@common/constants/Pagination'

export enum SortFields {
  LastUpdatedAt = 'lastModifiedAt',
  AZ09 = 'AZ09',
  ZA90 = 'ZA90',
  Name = 'name'
}

export enum Sort {
  DESC = 'DESC',
  ASC = 'ASC'
}

export interface PageQueryParams {
  page?: number
  size?: number
  searchTerm?: string
  filterIdentifier?: string
  filters?: FilterProperties
}

export type PageQueryParamsWithDefaults = RequiredPick<PageQueryParams, 'page' | 'size'>
export const PAGE_TEMPLATE_DEFAULT_PAGE_SIZE = 10
export const PAGE_TEMPLATE_DEFAULT_PAGE_INDEX = 1

export const usePageQueryParamOptions = (): UseQueryParamsOptions<PageQueryParamsWithDefaults> => {
  const { PL_NEW_PAGE_SIZE } = useFeatureFlags()

  return useQueryParamsOptions({
    page: PAGE_TEMPLATE_DEFAULT_PAGE_INDEX,
    size: PL_NEW_PAGE_SIZE ? COMMON_DEFAULT_PAGE_SIZE : PAGE_TEMPLATE_DEFAULT_PAGE_SIZE
  })
}
