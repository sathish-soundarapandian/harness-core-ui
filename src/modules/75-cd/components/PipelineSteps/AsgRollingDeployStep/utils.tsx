/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import * as Yup from 'yup'
import { getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'
import type { UseStringsReturn } from 'framework/strings'

export const minimumHealthyPercentageSchema = (
  getString: UseStringsReturn['getString']
): Yup.Schema<number | undefined> => {
  return Yup.lazy((value: number): any => {
    /* istanbul ignore else */
    if (getMultiTypeFromValue(value as any) === MultiTypeInputType.FIXED) {
      return Yup.mixed().test({
        test(): boolean | Yup.ValidationError {
          if (!value) {
            return true
          }
          if (value > 100) {
            return this.createError({ message: getString('cd.minimumHealthyPercentageMaxLimit') })
          }
          if (value < 0) {
            return this.createError({ message: getString('cd.minimumHealthyPercentageMinLimit') })
          }
          return true
        }
      })
    }
    return Yup.number()
  })
}

export const instanceWarmupSchema = (getString: UseStringsReturn['getString']): Yup.Schema<number | undefined> => {
  return Yup.lazy((value: number): any => {
    /* istanbul ignore else */
    if (getMultiTypeFromValue(value as any) === MultiTypeInputType.FIXED) {
      return Yup.mixed().test({
        test(): boolean | Yup.ValidationError {
          if (!value) {
            return true
          }
          if (value < 0) {
            return this.createError({ message: getString('cd.instanceWarmupError') })
          }
          return true
        }
      })
    }
    return Yup.number()
  })
}
