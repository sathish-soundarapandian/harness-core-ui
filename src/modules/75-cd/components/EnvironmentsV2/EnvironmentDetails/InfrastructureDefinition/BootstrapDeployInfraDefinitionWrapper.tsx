/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useMemo } from 'react'
import { useParams } from 'react-router-dom'
import { parse } from 'yaml'
import { defaultTo, merge, set } from 'lodash-es'
import produce from 'immer'

import type { InfrastructureConfig, InfrastructureDefinitionConfig } from 'services/cd-ng'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { DefaultPipeline } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import type {
  GetTemplateProps,
  GetTemplateResponse
} from 'framework/Templates/TemplateSelectorContext/useTemplateSelector'
import type { DeploymentStageElementConfig } from '@pipeline/utils/pipelineTypes'
import { DefaultNewStageId, DefaultNewStageName } from '@cd/components/Services/utils/ServiceUtils'
import { ServiceDeploymentType, StageType } from '@pipeline/utils/stageHelpers'
import { usePermission } from '@rbac/hooks/usePermission'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import { Scope } from '@common/interfaces/SecretsInterface'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import { InfrastructurePipelineProvider } from '@cd/context/InfrastructurePipelineContext'
import type { PipelineInfoConfig, TemplateLinkConfig } from 'services/pipeline-ng'
import { PipelineVariablesContextProvider } from '@pipeline/components/PipelineVariablesContext/PipelineVariablesContext'
import { DeployStageErrorProvider } from '@pipeline/context/StageErrorContext'
import { BootstrapDeployInfraDefinitionWithRef } from '@cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/BootstrapDeployInfraDefinition'

export interface InfraDefinitionWrapperRef {
  saveInfrastructure: () => void
}

interface WrapperProps {
  closeInfraDefinitionDetails: () => void
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  refetch: any
  scope: Scope
  selectedInfrastructure?: string
  environmentIdentifier: string
  stageDeploymentType?: ServiceDeploymentType
  stageCustomDeploymentData?: TemplateLinkConfig
  getTemplate?: (data: GetTemplateProps) => Promise<GetTemplateResponse>
  isDrawerView?: boolean
  setInfraSaveInProgress?: (data: boolean) => void
  handleInfrastructureUpdate?: (updatedInfrastructure: InfrastructureConfig) => void
  updatedInfra?: InfrastructureConfig
  isInfraUpdated?: boolean
}

export function BootstrapDeployInfraDefinitionWrapper(
  props: WrapperProps,
  infraDefinitionFormRef: React.ForwardedRef<InfraDefinitionWrapperRef>
) {
  const {
    closeInfraDefinitionDetails,
    refetch,
    selectedInfrastructure,
    environmentIdentifier,
    stageDeploymentType,
    stageCustomDeploymentData,
    getTemplate,
    scope,
    isDrawerView = false,
    setInfraSaveInProgress,
    handleInfrastructureUpdate,
    updatedInfra,
    isInfraUpdated
  } = props

  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()

  const infrastructureDefinition = useMemo(() => {
    return /* istanbul ignore next */ (parse(defaultTo(selectedInfrastructure, '{}')) as InfrastructureConfig)
      ?.infrastructureDefinition
  }, [selectedInfrastructure])

  const { type, spec, allowSimultaneousDeployments, deploymentType, identifier } = defaultTo(
    infrastructureDefinition,
    {}
  ) as InfrastructureDefinitionConfig

  const pipeline = React.useMemo(
    () =>
      produce({ ...DefaultPipeline }, draft => {
        set(
          draft,
          'stages[0].stage',
          merge({}, {} as DeploymentStageElementConfig, {
            name: DefaultNewStageName,
            identifier: DefaultNewStageId,
            type: StageType.DEPLOY,
            spec: {
              infrastructure: {
                infrastructureDefinition: {
                  ...(Boolean(type) && { type }),
                  ...(Boolean(spec) && { spec })
                },
                allowSimultaneousDeployments: Boolean(allowSimultaneousDeployments)
              },
              serviceConfig: {
                serviceDefinition: {
                  type: deploymentType
                }
              }
            }
          })
        )
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  )

  const [canEditInfrastructure] = usePermission(
    {
      resource: {
        resourceType: ResourceType.ENVIRONMENT,
        resourceIdentifier: environmentIdentifier
      },
      resourceScope: {
        accountIdentifier: accountId,
        ...(scope !== Scope.ACCOUNT && { orgIdentifier }),
        ...(scope === Scope.PROJECT && { projectIdentifier })
      },
      permissions: [PermissionIdentifier.EDIT_ENVIRONMENT]
    },
    [identifier]
  )

  return (
    <InfrastructurePipelineProvider
      queryParams={{ accountIdentifier: accountId, orgIdentifier, projectIdentifier }}
      initialValue={pipeline as PipelineInfoConfig}
      isReadOnly={!canEditInfrastructure}
      handleInfrastructureUpdate={handleInfrastructureUpdate}
    >
      <PipelineVariablesContextProvider pipeline={pipeline}>
        <DeployStageErrorProvider>
          <BootstrapDeployInfraDefinitionWithRef
            closeInfraDefinitionDetails={closeInfraDefinitionDetails}
            refetch={refetch}
            infrastructureDefinition={infrastructureDefinition}
            environmentIdentifier={environmentIdentifier}
            stageDeploymentType={(deploymentType as Partial<ServiceDeploymentType>) || stageDeploymentType}
            isReadOnly={!canEditInfrastructure}
            getTemplate={getTemplate}
            stageCustomDeploymentData={stageCustomDeploymentData}
            selectedInfrastructure={selectedInfrastructure}
            scope={scope}
            isDrawerView={isDrawerView}
            ref={infraDefinitionFormRef}
            setInfraSaveInProgress={setInfraSaveInProgress}
            handleInfrastructureUpdate={handleInfrastructureUpdate}
            isInfraUpdated={isInfraUpdated}
            updatedInfra={updatedInfra}
          />
        </DeployStageErrorProvider>
      </PipelineVariablesContextProvider>
    </InfrastructurePipelineProvider>
  )
}

export const BootstrapDeployInfraDefinitionWrapperWithRef = React.forwardRef(BootstrapDeployInfraDefinitionWrapper)
