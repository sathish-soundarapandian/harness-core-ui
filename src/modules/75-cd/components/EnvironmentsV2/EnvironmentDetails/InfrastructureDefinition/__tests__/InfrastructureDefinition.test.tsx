/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import * as cdNgServices from 'services/cd-ng'

import routes from '@common/RouteDefinitions'
import { environmentPathProps, modulePathProps, projectPathProps } from '@common/utils/routeUtils'
import { findDialogContainer, findPopoverContainer, TestWrapper } from '@common/utils/testUtils'

import InfrastructureDefinition from '../InfrastructureDefinition'

import mockInfrastructureList from './__mocks__/mockInfrastructureList.json'

jest.mock('../InfrastructureModal', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-infra-modal" />
}))

jest.mock('@common/pages/entityUsage/EntityUsage', () => ({
  __esModule: true,
  default: () => <div data-testid="mock-entity-setup-usage" />
}))

jest.mock(
  '@cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/BootstrapDeployInfraDefinitionWrapper',
  () => ({
    ...(jest.requireActual(
      '@cd/components/EnvironmentsV2/EnvironmentDetails/InfrastructureDefinition/BootstrapDeployInfraDefinitionWrapper'
    ) as any),
    BootstrapDeployInfraDefinitionWrapperWithRef: () => {
      return <div data-testid="mock-bootstrap-deploy-infra-definition" />
    }
  })
)

const showSuccessMock = jest.fn()
const showErrorMock = jest.fn()

jest.mock('@harness/uicore', () => ({
  ...jest.requireActual('@harness/uicore'),
  useToaster: () => ({
    showSuccess: showSuccessMock,
    showError: showErrorMock
  })
}))
jest.spyOn(cdNgServices, 'useGetSettingValue').mockImplementation(() => {
  return { data: { data: { value: 'false' } } } as any
})

describe('Infrastructure Definition tests', () => {
  test('renders loading spinner', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: true,
        error: undefined,
        data: undefined,
        refetch: jest.fn()
      } as any
    })

    render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    await waitFor(() => expect(screen.queryByText('Loading, please wait...')).toBeInTheDocument())
  })

  test('displays error message', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: false,
        error: {
          message: 'Failed to fetch'
        },
        data: undefined,
        refetch: jest.fn()
      } as any
    })

    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('displays no list, but shows 2 create buttons and opens modal on click ', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: false,
        error: undefined,
        data: {
          data: {
            content: []
          }
        },
        refetch: jest.fn()
      } as any
    })

    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()

    const createNewButtons = screen.queryAllByText('pipelineSteps.deploy.infrastructure.infraDefinition')
    expect(createNewButtons).toHaveLength(2)

    fireEvent.click(createNewButtons![0])
    await waitFor(() => expect(screen.getByTestId('mock-infra-modal')).toBeInTheDocument())
  })

  test('opens a modal on click of edit', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: false,
        error: undefined,
        data: mockInfrastructureList,
        refetch: jest.fn()
      } as any
    })

    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    const moreButton = container.querySelector('[data-icon="more"]')
    fireEvent.click(moreButton!)

    const menu = findPopoverContainer()
    expect(menu).toBeTruthy()

    const editButton = document.querySelector('[data-icon="edit"]')
    fireEvent.click(editButton!)
    await waitFor(() => expect(screen.getByTestId('mock-bootstrap-deploy-infra-definition')).toBeInTheDocument())
  })

  test('opens delete confirmation dialog and shows success toast on delete', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: false,
        error: undefined,
        data: mockInfrastructureList,
        refetch: jest.fn()
      } as any
    })

    jest.spyOn(cdNgServices, 'useDeleteInfrastructure').mockImplementation(() => {
      return {
        mutate: jest.fn(() =>
          Promise.resolve({
            status: 'SUCCESS'
          })
        ),
        cancel: jest.fn(),
        loading: false,
        error: null
      } as any
    })

    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    const moreButton = container.querySelector('[data-icon="more"]')
    fireEvent.click(moreButton!)

    const menu = findPopoverContainer()
    await waitFor(() => expect(menu).toBeTruthy())

    const deleteButton = document.querySelector('[data-icon="trash"]')
    fireEvent.click(deleteButton!)

    const dialog = findDialogContainer()
    await waitFor(() => expect(dialog).toBeTruthy())

    const confirmButton = dialog?.querySelector('.bp3-button-text')
    fireEvent.click(confirmButton!)

    await waitFor(() => expect(showSuccessMock).toHaveBeenCalled())
  })

  test('opens delete confirmation dialog and shows error toast on delete', async () => {
    jest.spyOn(cdNgServices, 'useGetInfrastructureList').mockImplementation(() => {
      return {
        loading: false,
        error: undefined,
        data: mockInfrastructureList,
        refetch: jest.fn()
      } as any
    })

    jest.spyOn(cdNgServices, 'useDeleteInfrastructure').mockImplementation(() => {
      return {
        mutate: jest.fn().mockRejectedValue(null),
        cancel: jest.fn(),
        loading: false,
        error: {
          message: 'Failed to delete'
        }
      } as any
    })

    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentDetails({
          ...projectPathProps,
          ...modulePathProps,
          ...environmentPathProps
        })}
        pathParams={{
          accountId: 'dummy',
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          module: 'cd',
          environmentIdentifier: 'Env_1'
        }}
        queryParams={{ sectionId: 'INFRASTRUCTURE' }}
      >
        <InfrastructureDefinition />
      </TestWrapper>
    )

    const moreButton = container.querySelector('[data-icon="more"]')
    fireEvent.click(moreButton!)

    const menu = findPopoverContainer()
    await waitFor(() => expect(menu).toBeTruthy())

    const deleteButton = document.querySelector('[data-icon="trash"]')
    fireEvent.click(deleteButton!)

    const dialog = findDialogContainer()
    await waitFor(() => expect(dialog).toBeTruthy())

    const confirmButton = dialog?.querySelector('.bp3-button-text')
    fireEvent.click(confirmButton!)

    await waitFor(() => expect(showErrorMock).toHaveBeenCalled())
  })
})
