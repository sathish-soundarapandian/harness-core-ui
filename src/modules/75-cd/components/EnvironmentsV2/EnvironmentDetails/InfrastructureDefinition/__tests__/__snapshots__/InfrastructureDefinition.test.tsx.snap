// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infrastructure Definition tests displays error message 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--padding-left-medium StyledProps--padding-right-medium"
  >
    <div
      class="StyledProps--font StyledProps--main PageError--pageError StyledProps--flex StyledProps--flex-align-center-center"
      style="width: 100%; height: 100%;"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
        style="align-items: center; width: 500px;"
      >
        <span
          class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
          icon="error"
        >
          <svg
            data-icon="error"
            height="32"
            viewBox="0 0 20 20"
            width="32"
          >
            <desc>
              error
            </desc>
            <path
              d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-align-center StyledProps--color StyledProps--color-red500"
        >
          We cannot perform your request at the moment. Please try again.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Infrastructure Definition tests displays no list, but shows 2 create buttons and opens modal on click  1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--padding-left-medium StyledProps--padding-right-medium"
  >
    <button
      aria-label="pipelineSteps.deploy.infrastructure.infraDefinition"
      class="bp3-button Button--button StyledProps--font StyledProps--main StyledProps--font StyledProps--font-weight-bold Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link Button--size Button--small"
      type="button"
    >
      <span
        class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
        icon="plus"
      >
        <svg
          data-icon="plus"
          height="16"
          viewBox="0 0 16 16"
          width="16"
        >
          <desc>
            plus
          </desc>
          <path
            d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="bp3-button-text"
      >
        pipelineSteps.deploy.infrastructure.infraDefinition
      </span>
    </button>
    <div
      class="Layout--vertical StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
      style="height: 70vh;"
    >
      <img
        alt="cd.infrastructure.noInfrastructureInEnvironment"
        src="test-file-stub"
      />
      <h2
        class="StyledProps--font StyledProps--font-h2 StyledProps--main StyledProps--padding-top-xxlarge StyledProps--margin-bottom-large"
        level="2"
      >
        cd.infrastructure.noInfrastructureInEnvironment
      </h2>
      <button
        aria-label="pipelineSteps.deploy.infrastructure.infraDefinition"
        class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-primary"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
          icon="plus"
        >
          <svg
            data-icon="plus"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              plus
            </desc>
            <path
              d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="bp3-button-text"
        >
          pipelineSteps.deploy.infrastructure.infraDefinition
        </span>
      </button>
    </div>
  </div>
</div>
`;
